src/zope/app/authentication/browser/loginform.py

```diff
@@ -13,30 +13,35 @@
 ##############################################################################
 """Login Form
 
-$Id$
 """
 from zope.authentication.interfaces import IUnauthenticatedPrincipal
 
 class LoginForm(object):
     """Mix-in class to implement login form logic"""
-    
+
+    context = None
+    request = None
+    unauthenticated = None
+    camefrom = None
+
     def __call__(self):
         request = self.request
         principal = request.principal
 
         unauthenticated = IUnauthenticatedPrincipal.providedBy(principal)
         self.unauthenticated = unauthenticated
-        
+
         camefrom = request.get('camefrom')
-        if isinstance(camefrom, list):
+        if isinstance(camefrom, list): # pragma: no cover
             # this can happen on python2.6, as it changed the
             # behaviour of cgi.FieldStorage a bit.
+            # XXX: Just Python 2.6 or later too? Tests don't produce this.
```

Comment:
I haven't the energy to verify, but presumably this happens when the HTTP request contains multiple query parameters with the same name -- you could try `GET /@@loginForm.html/?camefrom=foo&camefrom=bar` to test this corner case.

And I believe the behavior is the same for all Python versions >= 2.6.  (TBH I don't remember Python 2.5 behaving differently).